File: src/_cpp/TowerDefenseBase/scenes/gamegs/BoxMenu.cpp
#include "BoxMenu.h"
#include "GameGS.h"
#include "UserData.h"
#include "ml/SmartScene.h"
#include "scenes/itemshop/ItemShop.h"
#include "Tutorial.h"
NS_CC_BEGIN





BoxMenu::BoxMenu()
: _selectedItem( 0 )
, _isItemSelected( false )
{}

BoxMenu::~BoxMenu()
{}

bool BoxMenu::init( const std::string & xml )
{
	do
	{
		CC_BREAK_IF( !Menu::init() );
		CC_BREAK_IF( !init_machine() );

		load( xml );

		displayCountItems();

		return true;
	}
	while( false );
	return false;
}

bool BoxMenu::init_machine()
{
	add_state(state_close, std::bind(&BoxMenu::callback_close, this)).set_string_name("close");
	add_state(state_open, std::bind(&BoxMenu::callback_open, this)).set_string_name("open");
	add_state(state_wait, std::bind(&BoxMenu::callback_wait, this)).set_string_name("wait");

	add_event( event_close );
	add_event( event_open );
	add_event( event_wait );
	add_event( event_cancel );

	FiniteState::Machine::state( state_close ).add_transition( event_open, state_open );
	FiniteState::Machine::state( state_open ).add_transition( event_close, state_close );
	FiniteState::Machine::state( state_open ).add_transition( event_wait, state_wait );
	FiniteState::Machine::state( state_wait ).add_transition( event_close, state_close );
	FiniteState::Machine::state( state_wait ).add_transition( event_cancel, state_open );

	FiniteState::Machine::state( state_close ).add_onDeactivateCallBack( std::bind( &BoxMenu::close_deactivate, this ) );

	start( state_close );

	return true;
}

void BoxMenu::onEnter()
{
	Menu::onEnter();
	displayCountItems();
}

bool BoxMenu::isItemSelected(){
	return _isItemSelected;
}

ccMenuCallback BoxMenu::get_callback_by_description( const std::string & name )
{	
	auto close = [this]( Ref* )mutable{ push_event( event_close ); process(); };
	auto open = [this]( Ref* )mutable{ push_event( event_open ); process(); };
	
	auto item = [this]( Ref*, unsigned index )mutable 
	{		
		if( _selectedItem != index )
		{
			_isItemSelected = true;
			push_event( event_cancel );
			process();

			_selectedItem = index; 			
			push_event(event_wait);
			TutorialManager::shared().dispatch("boxmenu_item_did_selected");
		}
		else
		{
			push_event( event_cancel ); 
		}
		process();
	};
	
	auto itemshop = [this]( Ref* )mutable 
	{
		auto shop = ItemShop::create();
		if( shop )
		{
			SmartScene * scene = dynamic_cast<SmartScene*>(getScene());
			scene->pushLayer( shop, true );
		}
	};


	if( name == "close" )
		return std::bind( close, std::placeholders::_1 );
	else if( name == "open" )
		return std::bind( open, std::placeholders::_1 );
	else if( name == "item1" )
		return std::bind( item, std::placeholders::_1, 1 );
	else if( name == "item2" )
		return std::bind( item, std::placeholders::_1, 2 );
	else if (name == "item3")
		return std::bind(item, std::placeholders::_1, 3);    	
	else if( name == "itemshop" )
		return std::bind( itemshop, std::placeholders::_1 );
	else
		return NodeExt::get_callback_by_description( name );
}

bool BoxMenu::onTouchBegan( Touch* touch, Event* event )
{
	auto state = current_state().get_name();
	switch( state )
	{
		case state_close:
			_isItemSelected = false;
			return Menu::onTouchBegan( touch, event );
		case state_open:
		{
			bool touchedItem = Menu::onTouchBegan( touch, event );
			bool autoclose = strToBool( getParamCollection().get( "autoclose", "false" ) );
			if( touchedItem == false && autoclose )
			{
				push_event( event_close );
				process();
			}
			return touchedItem;
		}
		case state_wait:
		{
			bool touchedItem = Menu::onTouchBegan( touch, event );
			if( touchedItem == false )
			{
				if( createItem( touch->getLocation() ) )
				{
					bool autoclose = strToBool( getParamCollection().get( "autoclose", "false" ) );
					if( autoclose )
					{
						push_event( event_close );
						process();
					}
					else
					{
						push_event( event_cancel );
						process();
					}
					TutorialManager::shared().dispatch( "boxmenu_item_did_created" );
				}
			}
			return touchedItem;
		}
	}
	return Menu::onTouchBegan( touch, event );
}

void BoxMenu::close(){ 
	push_event(event_close);
	process();
}

bool BoxMenu::createItem( const Point & location )
{
	assert( _selectedItem != 0 );
	if( _selectedItem > 0 )
	{
		std::string items[] = {
			"_laser",
			"_ice",
			"_dynamit",
		};
		auto point = GameGS::getInstance()->getMainLayer()->convertToNodeSpace( location );
		auto& board = GameGS::getInstance()->getGameBoard();
		auto item = board.createBonusItem( point, "bonusitem" + items[_selectedItem-1] );
		if( item )
		{
			UserData::shared().bonusitem_sub( _selectedItem, 1 );
			displayCountItems();
		}
		return item != nullptr;
	}
	return false;
}

void BoxMenu::displayCountItems()
{
	auto display = [this]( int index )
	{
		auto menuitem = getChildByName<MenuItem*>( "item" + intToStr( index ) );
		auto label = menuitem->getChildByName<Label*>( "count" );
		int count = UserData::shared().bonusitem_count( index );
		label->setString( intToStr( count ) );
		menuitem->setEnabled( count > 0 );
	};
	display( 1 );
	display( 2 );
	display( 3 );
}

void BoxMenu::callback_open()
{
	runEvent( "open2" );
	_selectedItem = 0;
}

void BoxMenu::callback_close()
{
	runEvent( "close" );
	TutorialManager::shared().dispatch( "boxmenu_did_close" );
}

void BoxMenu::callback_wait()
{
	assert( _selectedItem != 0 );
	runEvent( "item" + intToStr( _selectedItem ) );
}

void BoxMenu::close_deactivate()
{
	runEvent( "open" );
	TutorialManager::shared().dispatch( "boxmenu_did_open" );
}


NS_CC_END

File: src/_eu/scenes/gamegs/BoxMenu.js
/******************************************************************************
 * Copyright 2014-2016 Vladimir Tolmachev
 * Copyright 2016 Visionarity AG
 * Vladimir Tolmachev and Visionarity AG have unlimited commercial
 * licenses for commercial use and customization
 *
 * Author: Vladimir Tolmachev (tolm_vl@hotmail.com)
 * Ported C++ to Javascript: Visionarity AG / Vladimir Tolmachev
 * Project: Island Defense (JS)
 * If you received the code not from the author, please contact us
 ******************************************************************************/

//Define namespace
var EU = EU || {};


EU.BoxMenu = cc.Menu.extend({
    __BoxMenu: true,

    state: Object.freeze({
        state_close: 0,
        state_open: 1,
        state_wait: 2
    }),
    event: Object.freeze({
        event_open: 0,
        event_close: 1,
        event_wait: 2,
        event_cancel: 3
    }),
    /** @type {Integer} */ _selectedItemIndex: null,
    /** @type {Boolean} */ _isItemSelected: null,
    ctor: function (/**@type {var} */ xml) {
        this._selectedItemIndex = 0;
        this._isItemSelected = false;
        this._super();
        this.initExt();
        this.init_machine();
        this.load_str(xml);
        this.displayCountItems();
        return true;
    },
    init_machine: function () {
        this.initMachine();
        this.add_state(this.state.state_close, this.callback_close.bind(this)).set_string_name("close");
        this.add_state(this.state.state_open, this.callback_open.bind(this)).set_string_name("open");
        this.add_state(this.state.state_wait, this.callback_wait.bind(this)).set_string_name("wait");

        this.add_event(this.event.event_open).set_string_name("open");
        this.add_event(this.event.event_close).set_string_name("close");
        this.add_event(this.event.event_wait).set_string_name("wait");
        this.add_event(this.event.event_cancel).set_string_name("cancel");

        this.state_tag(this.state.state_close).add_transition(this.event.event_open, this.state.state_open);
        this.state_tag(this.state.state_open).add_transition(this.event.event_close, this.state.state_close);
        this.state_tag(this.state.state_open).add_transition(this.event.event_wait, this.state.state_wait);
        this.state_tag(this.state.state_wait).add_transition(this.event.event_close, this.state.state_close);
        this.state_tag(this.state.state_wait).add_transition(this.event.event_cancel, this.state.state_open);

        this.state_tag(this.state.state_close).add_onDeactivateCallBack(this.close_deactivate.bind(this));

        this.start(this.state.state_close);

        return true;
    },
    onEnter: function () {
        this._super();
        this.displayCountItems();
    },
    isItemSelected: function () {
        return this._isItemSelected;
    },
    get_callback_by_description: function (/**@type {var} */ name) {
        var close = function () {
            this.push_event(this.event.event_close);
            this.process();
        };
        var open = function () {
            this.push_event(this.event.event_open);
            this.process();
        };

        var item = function (index) {
            if (this._selectedItemIndex != index) {
                this._isItemSelected = true;
                this.push_event(this.event.event_cancel);
                this.process();

                this._selectedItemIndex = index;
                this.push_event(this.event.event_wait);
                EU.TutorialManager.dispatch("boxmenu_item_did_selected");
            }
            else {
                this.push_event(this.event.event_cancel);
            }
            this.process();
        };

        var itemshop = function () {
            var shop = new EU.ItemShop();
            if (shop) {
                var scene = EU.Common.getSceneOfNode(this);
                scene.pushLayer(shop, true);
            }
        };


        if (name == "close")
            return close.bind(this);
        else if (name == "open")
            return open.bind(this);
        else if (name == "item1")
            return item.bind(this, 1);
        else if (name == "item2")
            return item.bind(this, 2);
        else if (name == "item3")
            return item.bind(this, 3);
        else if (name == "itemshop")
            return itemshop.bind(this);
        else
            return null;
    },
    _onTouchBegan: function (/**@type {Touch*} */ touch, /**@type {Event*} */ event) {
        var self = event.getCurrentTarget();
        var state = self.current_state().get_name();
        switch (state) {
            case self.state.state_close:
                self._isItemSelected = false;
                return cc.Menu.prototype._onTouchBegan( touch, event);
            case self.state.state_open:
            {
                var touchedItem = cc.Menu.prototype._onTouchBegan( touch, event);
                var autoclose = EU.Common.strToBool(self.getParamCollection().get("autoclose", "false"));
                if (touchedItem == false && autoclose) {
                    self.push_event(self.event.event_close);
                    self.process();
                }
                return touchedItem;
            }
            case self.state.state_wait:
            {
                var touchedItem = cc.Menu.prototype._onTouchBegan( touch, event);
                if (touchedItem == false) {
                    if (self.createItem(touch.getLocation())) {
                        var autoclose = EU.Common.strToBool(self.getParamCollection().get("autoclose", "false"));
                        if (autoclose) {
                            self.push_event(self.event.event_close);
                            self.process();
                        }
                        else {
                            self.push_event(self.event.event_cancel);
                            self.process();
                        }
                        EU.TutorialManager.dispatch("boxmenu_item_did_created");
                    }
                }
                return touchedItem;
            }
        }
        return cc.Menu.prototype._onTouchBegan.call( touch, event);
    },
    _onTouchMoved: function (/**@type {Touch*} */ touch, /**@type {Event*} */ event) {
        cc.Menu.prototype._onTouchMoved.call( event.getCurrentTarget(), touch, event);
    },
    _onTouchEnded: function (/**@type {Touch*} */ touch, /**@type {Event*} */ event) {
        cc.Menu.prototype._onTouchEnded.call( event.getCurrentTarget(), touch, event)
    },
    close: function () {
        this.push_event(this.event.event_close);
        this.process();
    },
    createItem: function (/**@type {var} */ location) {
        EU.assert(this._selectedItemIndex != 0);
        if (this._selectedItemIndex > 0) {
            var items = [
                "_laser",
                "_ice",
                "_dynamit"
            ];
            var point = EU.GameGSInstance.getMainLayer().convertToNodeSpace(location);
            var board = EU.GameGSInstance.getGameBoard();
            var item = board.createBonusItem(point, "bonusitem" + items[this._selectedItemIndex - 1]);
            if (item) {
                EU.UserData.bonusitem_sub(this._selectedItemIndex, 1);
                this.displayCountItems();
            }
            return item != null;
        }
        return false;
    },
    displayCountItems: function () {
        var self = this;
        var display = function (index) {
            var menuitem = self.getChildByName("item" + index);
            if (!(menuitem instanceof cc.MenuItem)) menuitem = null;

            var label = menuitem.getChildByName("count");
            if (!(label instanceof cc.LabelBMFont)) label = null;

            var count = EU.UserData.bonusitem_count(index);
            label.setString(count);
            menuitem.setEnabled(count > 0);
        };
        display(1);
        display(2);
        display(3);
    },
    callback_open: function () {
        this.runEvent("open2");
        this._selectedItemIndex = 0;
    },
    callback_close: function () {
        this.runEvent("close");
        EU.TutorialManager.dispatch("boxmenu_did_close");
    },
    callback_wait: function () {
        EU.assert(this._selectedItemIndex != 0);
        this.runEvent("item" + this._selectedItemIndex);
    },
    close_deactivate: function () {
        this.runEvent("open");
        EU.TutorialManager.dispatch("boxmenu_did_open");
    }
});

EU.NodeExt.call(EU.BoxMenu.prototype);
EU.Machine.call(EU.BoxMenu.prototype);
